crates/ra_hir_ty/src/method_resolution.rs

```diff
@@ -79,11 +86,45 @@ impl CrateImplDefs {
     }
 
     pub fn lookup_impl_defs_for_trait(&self, tr: TraitId) -> impl Iterator<Item = ImplId> + '_ {
-        self.impls_by_trait.get(&tr).into_iter().flatten().copied()
+        self.impls_by_trait
+            .get(&tr)
+            .into_iter()
+            .flat_map(|m| m.values().flat_map(|v| v.iter().copied()))
+    }
+
+    pub fn lookup_impl_defs_for_trait_and_ty(
```

Comment:
Do we need both methods public? If we do, than perhaps make self-ty non optional and make the caller write the match?

Reply:
Did the latter, how about now?